src/transformers/models/mask2former/modeling_mask2former.py

```diff
@@ -0,0 +1,2416 @@
+        self.is_train = config.general_config["is_train"]
+        self.use_auxiliary_loss = config.general_config["use_auxiliary_loss"]
+
+        self.decoder_norm = nn.LayerNorm(config.decoder_config["hidden_dim"])
+        self.num_feature_levels = 3
+
+        self.layers = nn.ModuleList(
+            [Mask2FormerTransformerDecoderLayer(config) for _ in range(config.decoder_config["decoder_layers"] - 1)]
+        )
+
+        self.class_embed = nn.Linear(config.decoder_config["hidden_dim"], config.general_config["num_classes"] + 1)
+        self.mask_embed = Mask2FormerMLPPredictionHead(
+            config.decoder_config["hidden_dim"],
+            config.decoder_config["hidden_dim"],
+            config.decoder_config["mask_feature_size"],
+            3,
+        )
+
+    def forward(
+        self,
+        multi_stage_features=None,
+        multi_stage_positional_embeddings=None,
+        mask_features=None,
+        output=None,
+        query_embeddings=None,
+        size_list=None,
+        output_attentions=None,
+    ):
+
+        intermediate_class_predictions = []
+        intermediate_mask_predictions = []
+
+        # Prediction heads on learnable query features
+        outputs_class, outputs_mask, attention_mask = self.forward_prediction_heads(
+            output, mask_features, attention_mask_target_size=size_list[0]
+        )
+        intermediate_class_predictions.append(outputs_class)
+        intermediate_mask_predictions.append(outputs_mask)
+
+        attentions = ()
+        for index, layer in enumerate(self.layers):
+            layer_outputs = layer(
+                index=index,
+                output=output,
+                multi_stage_features=multi_stage_features,
+                multi_stage_positional_embeddings=multi_stage_positional_embeddings,
+                attention_mask=attention_mask,
+                query_embeddings=query_embeddings,
+                output_attentions=output_attentions,
+            )
+
+            output = layer_outputs[0]
+            attentions += (layer_outputs[1:],)
+
+            outputs_class, outputs_mask, attention_mask = self.forward_prediction_heads(
+                output, mask_features, attention_mask_target_size=size_list[(index + 1) % self.num_feature_levels]
+            )
+
+            intermediate_class_predictions.append(outputs_class)
+            intermediate_mask_predictions.append(outputs_mask)
+
+        return Mask2FormerTransformerDecoderOutput(
+            prediction_masks=intermediate_mask_predictions[-1],
+            prediction_class=intermediate_class_predictions[-1],
+            auxiliary_predictions=self._get_aux_predictions(
+                intermediate_class_predictions, intermediate_mask_predictions
+            )
+            if self.use_auxiliary_loss
+            else None,
+            attentions=attentions,
+        )
+
+    def forward_prediction_heads(self, output, mask_features, attention_mask_target_size):
```

Comment:
To make it more aligned with our overall design, it perhaps makes more sense to make the decoder output the hidden states, and apply the predictions heads in the `xxxForUniversalSegmentation` model